pkg/cloudprovider/aws/launchtemplate.go

```diff
@@ -170,57 +153,38 @@ func (p *LaunchTemplateProvider) createLaunchTemplate(ctx context.Context, optio
 					},
 				},
 			}},
-			SecurityGroupIds: securityGroupIds,
-			UserData:         userData,
-			ImageId:          amiID,
+			SecurityGroupIds: aws.StringSlice(options.SecurityGroups),
+			UserData:         aws.String(options.UserData),
+			ImageId:          aws.String(options.AMIID),
 		},
 	})
 	if err != nil {
-		return nil, fmt.Errorf("creating launch template, %w", err)
+		return nil, err
 	}
 	zap.S().Debugf("Successfully created default launch template, %s", *output.LaunchTemplate.LaunchTemplateName)
 	return output.LaunchTemplate, nil
 }
 
-func (p *LaunchTemplateProvider) getSecurityGroupIds(ctx context.Context, clusterName string) ([]*string, error) {
-	securityGroupIds := []*string{}
-	securityGroups, err := p.securityGroupProvider.Get(ctx, clusterName)
+func (p *LaunchTemplateProvider) getSecurityGroupIds(ctx context.Context, provisioner *v1alpha1.Provisioner, constraints *Constraints) ([]string, error) {
+	securityGroupIds := []string{}
+	securityGroups, err := p.securityGroupProvider.Get(ctx, provisioner, constraints)
 	if err != nil {
-		return nil, err
+		return nil, fmt.Errorf("getting security group ids, %w", err)
 	}
 	for _, securityGroup := range securityGroups {
-		securityGroupIds = append(securityGroupIds, securityGroup.GroupId)
+		securityGroupIds = append(securityGroupIds, aws.StringValue(securityGroup.GroupId))
 	}
 	return securityGroupIds, nil
 }
 
-func (p *LaunchTemplateProvider) getAMIID(ctx context.Context, arch string) (*string, error) {
-	version, err := p.kubeServerVersion()
-	if err != nil {
-		return nil, fmt.Errorf("kube server version, %w", err)
-	}
-	paramOutput, err := p.ssm.GetParameterWithContext(ctx, &ssm.GetParameterInput{
-		Name: aws.String(fmt.Sprintf("/aws/service/bottlerocket/aws-k8s-%s/%s/latest/image_id", version, arch)),
-	})
-	if err != nil {
-		return nil, fmt.Errorf("getting ssm parameter, %w", err)
-	}
-	return paramOutput.Parameter.Value, nil
-}
-
-func (p *LaunchTemplateProvider) getUserData(options *launchTemplateOptions) (*string, error) {
+func (p *LaunchTemplateProvider) getUserData(provisioner *v1alpha1.Provisioner, constraints *Constraints) string {
 	t := template.Must(template.New("userData").Parse(bottlerocketUserData))
 	var userData bytes.Buffer
-	if err := t.Execute(&userData, options); err != nil {
-		return nil, err
-	}
-	return aws.String(base64.StdEncoding.EncodeToString(userData.Bytes())), nil
-}
-
-func (p *LaunchTemplateProvider) kubeServerVersion() (string, error) {
-	version, err := p.clientSet.Discovery().ServerVersion()
-	if err != nil {
-		return "", err
+	if err := t.Execute(&userData, struct {
+		Constraints *Constraints
+		Cluster     v1alpha1.ClusterSpec
+	}{constraints, *provisioner.Spec.Cluster}); err != nil {
+		panic(fmt.Sprintf("Parsing user data from %v, %v, %s", provisioner, constraints, err.Error()))
```

Comment:
Seems like if missed the validation and there is something wrong user has added to provisioner, we want to panic here? This will impact others too, if there are multiple provisioners?

Reply:
We rely on cluster being set in so many other places that you'd seg fault well before this line. I want to address this holistically at some point, but for now, we assume some invariants.